### PR TITLE
Model Average Optimizer

### PR DESCRIPTION
**Pull request by jinxin0924**
I have implemented model average optimizer and open a new pr since the [original one](https://github.com/tensorflow/tensorflow/pull/11581) is closed.

### CODE REVIEW

**Comment by jinxin0924** [comment]
@alextp, Could you please review this pr? It is quite similar to EASGD. Thanks a lot for your help in EASGD!

**Comment by yifeif** [comment]
Thanks @jinxin0924. Could you resolve the conflicts?

**Comment by jinxin0924** [comment]
@yifeif Conflicts have been resolved. Thanks!

File: tensorflow/contrib/opt/python/training/model_average_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    Args:
+      grads_and_vars: List of (gradient, variable) pairs as returned by
+        compute_gradients().
+      global_step: Optional Variable to increment by one after the
+        variables have been updated.
+      name: Optional name for the returned operation.  Default to the
+        name passed to the Optimizer constructor.
+
+    Returns:
+      A conditional 'Operation' that update both local and global varibales or
```
Comment (alextp) [inline tensorflow/contrib/opt/python/training/model_average_optimizer.py]: s/varibales/variables

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/opt/python/training/model_average_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  """
+
+  def __init__(self, worker_device):
+    """Create a new `ElasticAverageCustomGetter`.
+
+    Args:
+      worker_device: String.  Name of the `worker` job.
+    """
+    self._worker_device = worker_device
+    self._map = {}
```
[the remark this answers is not in the document]
Comment (jinxin0924) [reply tensorflow/contrib/opt/python/training/model_average_optimizer.py]: Rename it to self._local_2_global

**Comment by jinxin0924** [comment]
Hi @alextp , any more advice?

**Review by alextp — APPROVED** [review]

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by jinxin0924** [comment]
Hi @alextp , could you help me figure out why the Linux CPU Test failed? I have looked at the console output, and found:
"Build step 'Execute shell' marked build as failure
Unable to get pull request builder trigger!!"
What does it mean?
Thank you!

**Comment by jinxin0924** [comment]
@caisq, could you please help me figure out why the Linux CPU Test failed? It seems the failure is not caused by this pr.

**Comment by jinxin0924** [comment]
@tensorflow-jenkins test this please

**Comment by jinxin0924** [comment]
@caisq,  could you please help me figure out why the Linux CPU Test failed?

**Comment by ceballosiker** [comment]
Hi @jinxin0924 ,

I've been working with distributed tensorflow for a while and I found this contribution interesting. I've tried to make it work with a simple example on cifar-10, following this tutorial: https://www.tensorflow.org/tutorials/images/deep_cnn

The code I wrote works perfectly with tf.train.SyncReplicasOptimizer, however I can't make it work with your custom optimizer. I always get a TypeError: _true_getter() got multiple values for argument 'shape'. [Model Average Test](https://github.com/Acuratio/distributed-tensorflow-tests/blob/master/model_average_classifier.py)

This error was being triggered by tf.train.get_or_create_global_step() and tf.train.ExponentialMovingAverage(), so I removed them. However I still get the same error at the definition of the ModelAverageOptimizer which I can't remove, so I can't make it work.

Do you have any ideas as of why is this happening or how could I solve this? Any help would be much appreciated. Thanks!

**Comment by jinxin0924** [comment]
Hi @ceballosiker 
I am sorry it should be a bug. You can take a look at https://github.com/tensorflow/tensorflow/pull/19661#event-1826361323. It should be able to fix the problem.

**Comment by ceballosiker** [comment]
@jinxin0924 

Thanks for the quick reply. I followed the instructions on that thread and solved the issue by modifying the model_average_optimizer.py with your suggestions, maybe you should update the optimizer to include those changes?

However, now the chief worker is blocking just after starting the session. I'll try to figure out why this is happening and keep you updated.

Thanks again for your help!

**Comment by jinxin0924** [comment]
@ceballosiker Yes, I have pulled those changes and it has been merged!

**Pull request merged**